packages/flutter/lib/src/material/page.dart

```diff
@@ -95,11 +95,24 @@ mixin MaterialRouteTransitionMixin<T> on PageRoute<T> {
   @override
   String? get barrierLabel => null;
 
+  DelegatedTransitionBuilder? _delegatedTransition;
+
+  @override
+  DelegatedTransitionBuilder? get delegatedTransition => _delegatedTransition;
+
+  set delegatedTransition(DelegatedTransitionBuilder? newTransition) {
+    _delegatedTransition = newTransition;
+  }
+
   @override
   bool canTransitionTo(TransitionRoute<dynamic> nextRoute) {
-    // Don't perform outgoing animation if the next route is a fullscreen dialog.
-    return (nextRoute is MaterialRouteTransitionMixin && !nextRoute.fullscreenDialog)
-      || (nextRoute is CupertinoRouteTransitionMixin && !nextRoute.fullscreenDialog);
+    return ((nextRoute is! PageRoute<T>) || !nextRoute.fullscreenDialog)
+      && ((nextRoute is MaterialRouteTransitionMixin) || (nextRoute is ModalRoute<T> && nextRoute.delegatedTransition != null));
```

Comment:
Same comment as above, this should be documented or broken apart into bools with good names.